### PR TITLE
Share menu support

### LINKED ISSUE #15: Sharing is unreliable
10.13 and later enable a "Share" menu on the document. I've found this function to be unreliable with BSManagedDocument in at least two ways:

* If the document is unsaved (auto-saved), a "Failed other writing" error occurs
* If the document is saved, the first Share seems to work fine, but the second Share results in a hang

A workaround is to return `NO` from [`allowsDocumentSharing`](https://developer.apple.com/documentation/appkit/nsdocument/2902303-allowsdocumentsharing?language=objc), which disables the Share menu entirely. (In my test app, it doesn't actually remove Share from the File menu, but rather removes the individual Share actions.)

### PR DESCRIPTION
**Pull request by evanmiller**
This PR builds off of #17 and is designed to fix  #15. The purpose is to ensure that File > Share works as expected with unsaved and autosaved documents. I will keep it in draft status until #17 is reviewed, tested, and merged.

The Share function in NSDocument is very similar to File > Duplicate. Instead of defining new Save operators, these are treated as `NSAutosaveElsewhereOperation`s with a saved URL that is not the same as the `autosavedContentsFileURL`. So I've tried to set things up so that Share and Duplicate can use the same code paths. One small difference is that Share appears to autosave to a temporary folder, whereas Duplicate autosaves a copy to the regular Autosave folder. (This difference does not affect the logic and is just something I noticed while debugging.)

Previously, duplicate was implemented by setting the `writingBlock` to perform a simple file copy. I have moved this file-copy `writingBlock` logic out of `duplicateAndReturnError:` and into `writeSafelyToURL:` so anything that is not a regular autosave, but which executes an "autosave elsewhere", can take advantage of it. Share falls into this category.

In this PR I have `duplicateAndReturnError:` and `shareDocumentWithSharingService:completionHandler:` follow essentially the same logic:

1. Ensure a backing store exists by calling `managedObjectContext` (this is not true without #17)
2. Perform a synchronous autosave
3. Call `super`

The [Apple docs](https://developer.apple.com/documentation/appkit/nsdocument/2902309-sharedocumentwithsharingservice?language=objc) on `shareDocumentWithSharingService:completionHandler:` state that "If the document has never been saved, this method prompts the user to save it before proceeding". In practice it seems that an autosave counts as a save, so that an autosaved document can be directly shared without having to choose a location on disk first. This statement is not true of Sharing to iCloud drive, which requires first saving to disk, and then manually moving the document to your iCloud Drive.

For purposes of review it may be helpful to see the diff against my `persistent-containers` branch:

https://github.com/evanmiller/BSManagedDocument/compare/persistent-containers...share-menu

There is plenty to keep you busy with testing this feature, but I'd recommend using Share with something simple like Mail because the iCloud Drive workflow is fairly convoluted, at least by Apple standards.

### CODE REVIEW

**Comment by evanmiller** [comment]
Peeling off the Draft label, review at your leisure. I will note that I saw an error on this branch earlier today; could have been a code sync issue with my apps or something else. Anyway, because this PR changes some core code paths, extensive review and testing is warranted.

**Comment by evanmiller** [comment]
Would be great to have this reviewed; I have the persistent containers code running in production now, and plan to do the same with the Share branch next month. Thanks.

**Comment by jerrykrinock** [comment]
I merged it in with a new local yesterday and have been testing.  I shall be back soon.

File: BSManagedDocument.m [hunk 1]
```diff
@@ -1470,31 +1388,27 @@ - (NSDocument *)duplicateAndReturnError:(NSError **)outError;
...
+
+- (void)shareDocumentWithSharingService:(NSSharingService *)sharingService completionHandler:(void (^)(BOOL))completionHandler API_AVAILABLE(macos(10.13)) {
+    // Accessing the MOC ensures a backing store exists
+    (void)[self managedObjectContext];
     
-    self.writingBlock = writingBlock;
-    NSDocument *result = [super duplicateAndReturnError:outError];
-    self.writingBlock = nil;
+    // Make sure copy on disk is up-to-date
+    if (![self fakeSynchronousAutosaveAndReturnError:nil]) return;
```
Comment (evanmiller) [inline BSManagedDocument.m:1408]: I wonder whether this branch ought to call `completionHandler(NO)` to indicate failure to the caller.

(It's strange that the API does not provide a way to return an error.)

**Comment by jerrykrinock** [comment]
I've completed my review and testing of this pull request.  To make a long story short, there was a thorny issue that turned out to be in my code.  I did not test the Share function, however the code looks good and does not break anything in my apps.

Please make the additional change you suggested two days ago (run completionHander passing NO).  Then I shall pull and merge.

**Comment by evanmiller** [comment]
Okay, code updated per my suggestion. Ready for merge from my perspective.

**Pull request merged**